### ISSUE TITLE
Expand "select stars" without requiring a schema

### ISSUE THREAD

**Issue by jtalmi** [tobymao/sqlglot#1301]
A [recent analysis](https://select.dev/posts/should-you-use-ctes-in-snowflake) showed that on Snowflake, there can be performance advantages to expanding "select *" with specific column names.

This would look something like this: 
```
WITH users AS (
    SELECT *
    FROM users_table),
orders AS (
    SELECT *
    FROM orders_table)
SELECT users.id, users.name, orders.order_id, orders.total
FROM users
JOIN orders ON users.id = orders.user_id
```
would get translated into:
```
WITH users AS (
    SELECT id, name
    FROM users_table),
orders AS (
    SELECT user_id, order_id, total
    FROM orders_table)
SELECT users.id, users.name, orders.order_id, orders.total
FROM users
JOIN orders ON users.id = orders.user_id
```

It should be possible to do this without a schema in most cases. I wasn't able to see an optimization that does this but I did see snippets that might be useful, e.g. the `_expand_stars` method, maybe qualifying column names.

I'd be happy to add this functionality if this issue is approved.

**Comment by GeorgeSittas** [comment]
Right now SQLGlot optimizes this query into the following:

```python
>>> from sqlglot.optimizer import optimize
>>> print(optimize(
... """
... WITH users AS (
...     SELECT *
...     FROM users_table),
... orders AS (
...     SELECT *
...     FROM orders_table)
... SELECT users.id, users.name, orders.order_id, orders.total
... FROM users
... JOIN orders ON users.id = orders.user_id
... """).sql(pretty=True)
... )
SELECT
  "users_table"."id" AS "id",
  "users_table"."name" AS "name",
  "orders_table"."order_id" AS "order_id",
  "orders_table"."total" AS "total"
FROM "users_table" AS "users_table"
JOIN "orders_table" AS "orders_table"
  ON "users_table"."id" = "orders_table"."user_id"
```

As you can see, it can already expand stars. Is this what you were suggesting or did I miss something?

**Comment by GeorgeSittas** [comment]
@tobymao what about using all the optimizer rules except for `merge_subqueries`, which is responsible for merging CTEs?

**Comment by tobymao** [comment]
i think that exists already with leave_tables_isolated

**Comment by jtalmi** [comment]
Yeah I was able to achieve this by passing in that kwarg. I dropped canonicalize as well since Snowflake doesn't like quoting. Thanks!

By the way, I built a [streamlit app](https://sql-optimizer.streamlit.app/) to showcase the various optimizations. I made it after I implemented expand_stars myself, but figured it would be cool to be able to batch together different optimizations and see what happens. Let me know what you think!
